loopout.go

```diff
@@ -1113,23 +1135,92 @@ func (s *loopOutSwap) waitForHtlcSpendConfirmed(globalCtx context.Context,
 		// Some time after start or after arrival of a new block, try
 		// to spend again.
 		case <-timerChan:
-			err := spendFunc()
-			if err != nil {
-				return nil, err
-			}
+			if s.isTaproot() {
+				// When using taproot HTLCs we're pushing the
+				// preimage before attempting to sweep. This
+				// way the server will know that the swap will
+				// go through and we'll be able to MuSig2
+				// cosign our sweep transaction. In the worst
+				// case if the server is uncooperative for any
+				// reason we can still sweep using scriptpath
+				// spend.
+				err = s.setStatePreimageRevealed(ctx)
+				if err != nil {
+					return nil, err
+				}
 
-			// If the result of our spend func was that the swap
-			// has reached a final state, then we return nil spend
-			// details, because there is no further action required
-			// for this swap.
-			if s.state.Type() != loopdb.StateTypePending {
-				return nil, nil
-			}
+				if !paymentComplete {
+					// Push the preimage for as long as the
+					// server is able to settle the swap
+					// invoice. So that we can continue
+					// with the MuSig2 sweep afterwards.
+					s.pushPreimage(ctx)
+				}
+
+				// Now attempt to publish a MuSig2 sweep txn.
+				// Only attempt at most maxMusigSweepRetires
+				// times to still leave time for an emergency
+				// script path sweep.
+				if musigSweepTryCount < maxMusigSweepRetires {
```

Comment:
Compared to the old flow, this does give the server extra 100s to cheat us. Since mining 20 blocks within that time is extremely unlikely so it doesn't matter. Only something to be noted if we decide one day to change. the value of `MinLoopOutPreimageRevealDelta`.

Reply:
Was curious about this so I looked it up. So the probability of finding k blocks in a row follows a Poisson[1] distribution, assuming 10 minute block time: `P(k, t) = (t/10)^k * e^(-t/10) / k!)`, where `t` is minutes, in this case it's 1.5 (100/60) and `k` is 20 (blocks in a row), which gives us `P(20, 1.5) = 0.15^20 * e^-0.15 / 20!`, so in Python `math.pow(0.15, 20) * math.pow(math.e, -0.15) / math.factorial(20)` which gives us  `1.176403727573749e-35`. Hope I'm not misunderstanding how to calculate :)

[1] https://en.wikipedia.org/wiki/Poisson_distribution